### PR TITLE
fixed nf to work with negative number

### LINKED ISSUE #7046: `nf()` produces problematic string-formatting of negative numbers
### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [ ] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [X] Utilities
- [ ] WebGL
- [ ] Build process
- [ ] Unit testing
- [ ] Internationalization
- [ ] Friendly errors
- [ ] Other (specify if possible)

### p5.js version

1.9.3

### Web browser and version

Chrome 124.0.6367.208

### Operating system

macOS 14.4.1

### Steps:
1. Call `nf()` with a negative number and extra left padding.
2. Display the formatted number.

### Snippet:

```js
function setup() {
  createCanvas(400, 400);

  background(220);

  // Format a negative number with
  // extra left padding.
  let num = nf(-123, 5);

  // Style the text.
  textAlign(CENTER, CENTER);
  textSize(32);

  // Display the text.
  text(num, 200, 200); // 0-123
}
```

Here's the [sketch](https://editor.p5js.org/mcintyre/sketches/XmxqHk65M) for reference.

After reviewing #5710, it may be worth revisiting the string utility functions as part of p5.js 2.0.

### PR DESCRIPTION
**Pull request by Akhilbisht798**
Resolves #[Add issue number here]

 Changes:

resolve #7046

 Screenshots of the change:

#### PR Checklist

Screenshot from 2024-05-16 21-49-43

- [x] `npm run lint` passes
- [ ] [Inline documentation] is included / updated
- [ ] [Unit tests] are included / updated

[Inline documentation]: https://github.com/processing/p5.js/blob/main/contributor_docs/inline_documentation.md
[Unit tests]: https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests

### CODE REVIEW

**Review by perminder-17 — CHANGES_REQUESTED** [review]
Firstly, I want to apologize for the delayed review, @Akhilbisht798. Due to the maintainers being occupied with the version update, I’ll be taking over the review of your work.

I appreciate your patience and the effort you’ve put into this task. Your work is mostly accurate, but there is one issue with the rounding of numbers after the decimal point that may need to be addressed.

For example, if we use `let num = nf(123.58, 1,1);`, the expected output should be `123.6`, as it involves rounding the numbers after the decimal point. However, in your implementation, it currently results in `123.5` without rounding off correctly. Do you think we could adjust this?

**Comment by Akhilbisht798** [comment]
hey @perminder-17 I can work on this.

**Comment by perminder-17** [comment]
Thanks, @Akhilbisht798, for your great work so far, and I apologize again for the delay—I hope you don’t mind. I’d like to invite the utility stewards, @limzykenneth, @glopzel, and @davepagurek and @nickmcintyre  to share their thoughts as well.

I noticed that @limzykenneth has already opened a discussion on this issue (https://github.com/processing/p5.js/issues/7046#issuecomment-2200266696).

In short, should we be writing `let num = nf(-123, 4);` where 4 represents the number of digits to include to the left? Currently, the output produced after solving by @Akhilbisht798  is -0123, but if we count the minus sign (-) as a character, the output would be -123. The question is: should we consider the minus sign in the digit count? 

From my opinion I go with **not** counting minus sign(-). 

I’d appreciate everyone’s input on this discussion that @limzykenneth started. Also, if you have any other suggestions regarding nf(), please feel free to share. (To be honest, I haven’t used nf() much myself, so I’m a bit uncertain about it.)

**Comment by limzykenneth** [comment]
I think we can possibly go with not counting the minus sign, which is also the behaviour of Processing. For those who needed the numbers to line up, `nfs()` can solve that.

File: src/utilities/string_functions.js [hunk 1]
```diff
@@ -295,20 +295,23 @@ p5.prototype.nf = function(nums, left, right) {
...
   if (typeof right === 'undefined') {
     leftPart = leftPart.padStart(left, '0');
-    return rightPart ? leftPart + '.' + rightPart : leftPart;
+    let result = rightPart ? leftPart + '.' + rightPart : leftPart;
+    return isNegative ? '-' + result : result;
   } else {
-    let roundedOff = num.toFixed(right);
-    [leftPart, rightPart] = roundedOff.toString().split('.');
+    let roundedNum = Math.round(num * Math.pow(10, right));
+    roundedNum = roundedNum / Math.pow(10, right);
```
Comment (perminder-17) [inline src/utilities/string_functions.js:309]: Can we roundooff the num by simply doing this:
```suggestion
    let roundedNum = num.toFixed(right);
```

Let me know if it works for you @Akhilbisht798 .

File: src/utilities/string_functions.js [hunk 2]
```diff
@@ -295,20 +295,23 @@ p5.prototype.nf = function(nums, left, right) {
...
+    [leftPart, rightPart] = roundedNum.toString().split('.');
     leftPart = leftPart.padStart(left, '0');
-    if(typeof rightPart === 'undefined'){
-      return leftPart;
-    }else{
-      return leftPart + '.' + rightPart;
-    }
+    let result = typeof rightPart === 'undefined' ?
+      leftPart : leftPart + '.' + rightPart.padEnd(right, '0');
```
Comment (perminder-17) [inline src/utilities/string_functions.js:313]: Was there a need to write `.padEnd`?
```suggestion
      leftPart : leftPart + '.' + rightPart;
```

can we simply add this? What you think?

File: src/utilities/string_functions.js [hunk 3]
```diff
@@ -295,20 +295,23 @@ p5.prototype.nf = function(nums, left, right) {
...
+    let roundedNum = Math.round(num * Math.pow(10, right));
+    roundedNum = roundedNum / Math.pow(10, right);
+    [leftPart, rightPart] = roundedNum.toString().split('.');
     leftPart = leftPart.padStart(left, '0');
-    if(typeof rightPart === 'undefined'){
-      return leftPart;
-    }else{
-      return leftPart + '.' + rightPart;
-    }
+    let result = typeof rightPart === 'undefined' ?
```
Comment (perminder-17) [inline src/utilities/string_functions.js:312]: Do you think @Akhilbisht798 that using ternary operator for result could be a bit complicated for the other contributors to understand in a first look. Can we use if and else block for this part only? Do you agree, or maybe you have any other thoughts? Please do let me know.

**Review by perminder-17 — COMMENTED** [review]
Everything looks good, Thanks @limzykenneth for your response..just some minor cleanups then we are good to merge. Thanks @Akhilbisht798 for your hardwork and patience :)

File: src/utilities/string_functions.js (hunk 1 above)
Comment (Akhilbisht798) [reply src/utilities/string_functions.js:309]: yes we can, previous commit was using it. but as you have mentioned that 

> For example, if we use let num = nf(123.58, 1,1);, the expected output should be 123.6, as it involves rounding the numbers after the decimal point. However, in your implementation, it currently results in 123.5 without rounding off correctly. Do you think we could adjust this?


This is was not working with `num.toFixed(right);`. But if you want I can change it back.

**Comment by ashish1729** [comment]
do we need to update some test + reference ?

File: src/utilities/string_functions.js (hunk 2 above)
Comment (Akhilbisht798) [reply src/utilities/string_functions.js:313]: So this add padding of 0 at the right decimal place if the right is more than the digit in the right. 
if we doNf with right of 2 in 123.5 it will pad 123.50.
I guess it would be great. But tell me known what you think about this. I can change if it is not required.

File: src/utilities/string_functions.js (hunk 3 above)
Comment (Akhilbisht798) [reply src/utilities/string_functions.js:312]: I have no problem in using if else block for this statement if that increase the look. Let me known what you think about the other changes and I will commit the changes of this with them.

Comment (ashish1729) [reply src/utilities/string_functions.js:312]: just a thought...  

if we want to reduce the usage of ternary operator it can be mentioned in the contributor guideline somewhere

or if we have strong opinions about it, it should be part of some lint system to give warning / error

**Comment by perminder-17** [comment]
Hmm..I realize I made an error earlier. When I reviewed the code again, I found that it wasn’t rounding the numbers as I initially thought, and the mistake was on my side. I sincerely apologize for any confusion caused. It seems your previous commit was indeed correct. I hadn’t thoroughly checked the code and was only focusing on the testing, which led me to overlook this. After testing it again now, everything appears to be working perfectly. 

here's the build of your previous commit:
https://editor.p5js.org/aman12345/sketches/Pq9D8HM0h

We can revert to the previous commit, as it looks all good to me. I’m really sorry for any inconvenience caused, and then we could merge this. Really thankful for your patience and hard work @Akhilbisht798.

File: src/utilities/string_functions.js (hunk 2 above)
Comment (perminder-17) [reply src/utilities/string_functions.js:313]: Right, when we will use `num.toFixed(right)` to roundoff the numbers, I don't think we would actually need to add padding of 0 to the right.

**Comment by perminder-17** [comment]
> do we need to update some test + reference ?

Yes, adding tests would be a great idea. I believe we already have a good reference sketch with a detailed explanation in the code. If @Akhilbisht798 has some time, we could also add a test for negative numbers, which is currently causing issues. You can check it here: 
here: https://github.com/processing/p5.js/blob/7d393ec4215ce20f7bc3178b779e0ec67689f928/test/unit/utilities/string_functions.js#L79 

we could probably add the same way as it is written here ;)

File: src/utilities/string_functions.js (hunk 1 above)
Comment (perminder-17) [reply src/utilities/string_functions.js:309]: Were you also able to reproduce the same error of rounding off with `num.toFixed(right);`. I am currently testing and here's the build and it works well.

https://editor.p5js.org/aman12345/sketches/Pq9D8HM0h

File: src/utilities/string_functions.js (hunk 3 above)
Comment (perminder-17) [reply src/utilities/string_functions.js:312]: > just a thought...
> 
> if we want to reduce the usage of ternary operator it can be mentioned in the contributor guideline somewhere
> 
> or if we have strong opinions about it, it should be part of some lint system to give warning / error

It's not about reducing the use of ternary operators. In this case, I was just considering whether using if and else blocks might make the code more readable, especially when checking the type. If it's undefined, we perform one operation, and for everything else, we do something different—all in a single line. I'm open to discussion and would love to hear your thoughts on this. Otherwise ternary operator also works for me. Any feedback @Akhilbisht798 and @ashish1729 ? 

```
let result = typeof rightPart === 'undefined' ? leftPart : leftPart + '.' + rightPart;
```

**Comment by Akhilbisht798** [comment]
@perminder-17 It is not a problem. I will revert back to the previous commit and will add the test case for this one also.

**Comment by Akhilbisht798** [comment]
@perminder-17 all done you can let me known.

**Review by perminder-17 — APPROVED** [review]
Looks good @Akhilbisht798 . Thanks for your hard work:)

**Comment by ashish1729** [comment]
should update the reference for negative numbers

I can see myself using this function for having a fixed-width display... but the way it is implemented negative numbers have more width due to the extra minus sign.

can be taken up as a separate task to unblock the merging of this PR

**Comment by perminder-17** [comment]
> should update the reference for negative numbers

Yes, we could consider adding support for negative numbers as well. However, since the issue specifically addresses the bug 'nf() produces problematic string formatting for negative numbers,' this PR resolves that issue correctly. It might be a good idea to handle any additional improvements in a separate task.

**Comment by ashish1729** [comment]
oh by reference I meant documentation, the reference page of the website ..

**Comment by perminder-17** [comment]
> oh by reference I meant documentation, the reference page of the website ..

Yep, I was imagining the same. Sorry... I will make the above comment more clearer. Haha😅

**Comment by ashish1729** [comment]
then why is documentation considered additional improvement ? shouldn't it be in sync with any update in the code?


the PR on top has 3 checkboxes, which is included in every PR
- npm run lint passes
- [Inline documentation](https://github.com/processing/p5.js/blob/main/contributor_docs/inline_documentation.md) is included / updated
- [Unit tests](https://github.com/processing/p5.js/tree/main/contributor_docs#unit-tests) are included / updated


It was my understanding that documentation should be updated to check the 2nd point

**Comment by perminder-17** [comment]
That's a good point. However, since the PR has been open for five months and @Akhilbisht798 has been very patient and kind in making updates and adding tests, requesting more changes one by one might not be ideal. @Akhilbisht798, if you're interested, it would be great if you could add the documentation for negative numbers, but if you're busy, I'd be happy to work on it and submit a PR. Again thanks a lot for your hard work and patience.

**Comment by ashish1729** [comment]
I am also trying to learn how thing works, so thank you for explaining.

**Comment by Akhilbisht798** [comment]
@perminder-17 No problem, I can add documentation changes. 
but I personally think that the inline documentation for this function is fine and explain things perfectly and to just a changes for negative number won't be that big of deal and won't be changing the functionality in a big way

**Comment by perminder-17** [comment]
Right, the current documentation is also very good no doubt. But as @ashish1729 said, 

> I can see myself using this function for having a fixed-width display... but the way it is implemented negative numbers have more width due to the extra minus sign.

Also, one more reason could be, we are not counting the minus sign, so we could also add a line of comment describing it in the code with an example sketch where it shows a negative  number. 

What you think?

**Comment by Akhilbisht798** [comment]
ya no problem i can do that.

File: src/utilities/string_functions.js [hunk 4]
```diff
@@ -222,6 +222,9 @@ p5.prototype.matchAll = function(str, reg) {
  * then unused decimal places will be set to 0. For example, calling
  * `nf(123.45, 4, 3)` returns the string `'0123.450'`.
  *
+ * When the number is negative, for example, calling `nf(-123.45, 5, 2)`
```
Comment (perminder-17) [inline src/utilities/string_functions.js:225]: Thank you for adding that. Just one last thought: could you also modify the reference example so that it appears like this?

image

Since we are not counting the minus sign as an extra digit, negative numbers may have greater width due to the additional minus sign. Therefore, we could show users that negative numbers occupy more space as we are not counting it.

Comment (Akhilbisht798) [reply src/utilities/string_functions.js:225]: ya no problem.

**Comment by Akhilbisht798** [comment]
hey @perminder-17 let me known

**Review by perminder-17 — APPROVED** [review]
Looks good to me. Thanks for your work and your patience.
